mobilecoind/src/service.rs

```diff
@@ -4184,6 +4292,226 @@ mod test {
+        let (mut ledger_db, mobilecoind_db, client, _server, _server_conn_manager) =
+            get_testing_environment(
+                BLOCK_VERSION,
+                3,
+                &[
+                    swap_originator.default_subaddress(),
+                    swap_counterparty.default_subaddress(),
+                ],
+                &[],
+                logger.clone(),
+                &mut rng,
+            );
+
+        // Add a block with a non-MOB token ID.
+        add_block_to_ledger(
+            &mut ledger_db,
+            BlockVersion::MAX,
+            &[
+                AccountKey::random(&mut rng).default_subaddress(),
+                AccountKey::random(&mut rng).default_subaddress(),
+                swap_originator.default_subaddress(),
+                swap_counterparty.default_subaddress(),
+            ],
+            Amount::new(1_000_000_000_000, TokenId::from(1)),
+            &[KeyImage::from(101)],
+            &mut rng,
+        )
+        .unwrap();
+
+        // Insert into database.
+        let originator_monitor_id = mobilecoind_db.add_monitor(&swap_originator_data).unwrap();
+        let counterparty_monitor_id = mobilecoind_db.add_monitor(&swap_counterparty_data).unwrap();
+
+        // Allow the new monitor to process the ledger.
+        wait_for_monitors(&mobilecoind_db, &ledger_db, &logger);
+
+        // Get list of unspent tx outs
+        let utxos = mobilecoind_db
+            .get_utxos_for_subaddress(&originator_monitor_id, 0)
+            .unwrap();
+        assert!(!utxos.is_empty());
+
+        let offered_input = utxos
+            .iter()
+            .filter(|utxo| utxo.token_id == *Mob::ID)
+            .map(api::UnspentTxOut::from)
+            .next()
+            .unwrap();
+        let offered_value = offered_input.value;
+
+        // Generate a swap.
+        let generate_swap_response = {
+            let mut request = api::GenerateSwapRequest::new();
+            request.set_sender_monitor_id(originator_monitor_id.to_vec());
+            request.set_change_subaddress(0);
+            request.set_input(offered_input);
+            request.set_allow_partial_fill(true);
+            request.set_counter_value(123);
+            request.set_counter_token_id(1);
+            request.set_minimum_fill_value(10);
+            client.generate_swap(&request).unwrap()
+        };
+
+        // Now we will try to build a transaction that incorporates the swap.
+        // The counterparty needs to supply eUSD.
+        // Get list of unspent tx outs
+        let utxos = mobilecoind_db
+            .get_utxos_for_subaddress(&counterparty_monitor_id, 0)
+            .unwrap();
+        assert!(!utxos.is_empty());
+
+        // We will try to take half of the offered input.
+        let mut sci_for_tx = api::SciForTx::new();
+        sci_for_tx.set_sci(generate_swap_response.get_sci().clone());
+        sci_for_tx.set_partial_fill_value(offered_value / 4);
```

Comment:
Doesn't this imply taking a quarter?

Reply:
yeah i have to update the comment, thanks